test: introduce ch-benchmark

nanderstabel opened the pull request:
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

To introduce CH-benCHmark.

In this PR:
* TPC-C table events in debezium json format are included. They consist of reads, creates, deletes and inserts
* The creates, deletes and inserts are modeled after the [delivery](https://www.tpc.org/tpc_documents_current_versions/pdf/[tpc-c_v5.11.0.pdf](https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf#page=40)#page=40) and [new order](https://www.tpc.org/tpc_documents_current_versions/pdf/[tpc-c_v5.11.0.pdf](https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf#page=40)#page=40) transactions as described in the official TPC BENCHMARK™ C documentation.
* The e2e test can be found in `e2e_test/ch-benchmark/ch_benchmark.slt`. It creates materialized sources from each TPC-C json files and runs two CH-benCHmark queries (correctly). This benchmark test is currently added to the e2e-source-test script so that it can utilize the `prepare_ci_kafka.sh` script. At a later stage though it should be transferred to the benchmark CI workflow. 

## Notes
* TPC-C tables are explicitly prefixed `tpcc_`. This is in order to avoid confusion since both TPC-H and TPC-C have a customers and order(s) table.
* In the original CH-benCHmark queries the mathematical function `MOD()` is used. They are currently replaced by the modulo operator `%`.
* Table name prefix of `tpcc_stock` table is changed from `s_` to `st_` in order to distinguish it from the `supplier` table name prefix (in tpc-h).


## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

BugenZhao commented:
Seems like there're some issues to be resolved. Would you please open a tracking issue for this? Like #1380.

neverchanje reviewed (approved):
Looks like only q13 and q16 were passed?

nanderstabel commented:
> Looks like only q13 and q16 were passed?

Most queries do pass (run without panic) but do not produce output due to limited amount of data in the TPC-C tables in the postgres database container. q13 and q16 run successfully as 'statements', but the data in this PR are only READS, not a whole bunch of UPDATES/DELETES/INSERTS yet. The updates in the simple transaction query I added in `ci/ch-benchmark/transactions/example_update.sql` are actually not even parsed (and thus skipped) by RisingWave. I have added these two queries more to demonstrate that the whole pipeline is functional. The actual content of the queries is not that important right now I guess. The logical steps regarding this benchmark imo are these:
1. Build the postgres->debezium->kafka->RisingWave pipeline (this PR)
2. Add this pipeline to the CI benchmark pipeline (Sumit will help me with this)
3. Add a whole bunch of transactions that include UPDATES/DELETES/INSERTS. We/I need to decide how complex to make these transactions. Probably the best way to go is to start off with some very simple transaction to catch the first bugs. At a later stage we can decide to add the actual 'official' TPC-C transactions that consist of multiple inserts/updates and randomly generated data (more details: https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf#page=28).
4. After these steps are completed it makes sense to add more of the CH-benCHmark queries. I already 'translated' these queries locally. The main goal of this step would be to curate specific input data for each of those queries to produce actual output. However, like @fuyufjh mentioned: priority is to catch any bugs that arise just by ingesting update/delete input from kafka. This is unrelated to the actual CH-benCHmark queries. Only once most of those bugs are found and fixed it makes sense to add more queries and look at their behavior. 

A bit lengthy response 😂 but it also helped to structure my thought. I will create a tracking issue for this 👍

neverchanje commented:
I see, but you don't need to actually start Postgres and a Debezium server, if the data has been pre-generated.
You can run Postgres+Debezium+Kafka ([Redpanda](https://docs.redpanda.com/docs/quickstart/quick-start-linux/) is recommended) locally, and ingest the prepared dataset into Pg. After a while, you will get a bunch of records flushed to Kafka. Then, you can output the records in Kafka to a single JSON file. 

```bash
rpk topic consume -f '%v\n' --brokers 127.0.0.1:55971
```

Finally, the JSON file can be used as the test input for CI.

nanderstabel commented:
Yes, I agree that would be a better way to go about this (and to not change the current workflow) 🤔 Perhaps that the updates I made to docker-compose.yml can still act as a useful tool to generate records though. But I indeed think that step 2. can be skipped at least for now. So, I will continue with step 3. since that has the highest priority atm 👍

neverchanje commented:
Yeah, just removing debezium and postgres for now in the docker-compose file. Add them back if really needed.

fuyufjh commented on scripts/source/test_data/tpcc_warehouse.1:
```diff
@@ -0,0 +1,2 @@
+{"schema":{"type":"struct","fields":[{"type":"struct","fields":[{"type":"int32","optional":false,"field":"w_id"},{"type":"string","optional":true,"field":"w_name"},{"type":"string","optional":true,"field":"w_street_1"},{"type":"string","optional":true,"field":"w_street_2"},{"type":"string","optional":true,"field":"w_city"},{"type":"string","optional":true,"field":"w_state"},{"type":"string","opti ...
+^C
```
Is it intentional?

fuyufjh commented:
> I will create a tracking issue for this 👍

Have you create one? If so, please link it here. BTW, how about close (or update) this issue: https://github.com/risingwavelabs/risingwave/issues/5190

fuyufjh commented on e2e_test/ch-benchmark/create_sources.slt.part:
```diff
@@ -0,0 +1,181 @@
+statement ok
+CREATE MATERIALIZED SOURCE tpcc_customer (
```
> TPC-C tables are explicitly prefixed tpcc_. This is in order to avoid confusion since both TPC-H and TPC-C have a customers and order(s) table.


The renaming doesn't seem to be necessary because we create & drop tables for each test cases. Correct me if wrong

nanderstabel replied:
It is because ch-benchmark also utilizes some TPC-H tables. I include those in `e2e_test/ch-benchmark/ch_benchmark.slt` like this `include ../tpch/create_tables.slt.part`. `../tpch/create_tables.slt.part` creates all TPC-H tables so also those with duplicate names (even though they're not actually used in the ch-benchmark. 

So in short, in the current situation both `customer` and `orders` tables are created (TPC-H) as well as `tpcc_customer` and `tpcc_orders` of which only the latter two are actually used.

fuyufjh reviewed (approved):
Generally LGTM

nanderstabel commented:
> 

Yes, will update 👍

nanderstabel commented:
> > I will create a tracking issue for this 👍
> 
> Have you create one? If so, please link it here. BTW, how about close (or update) this issue: #5190

Yes, will update 👍
